File: lambda_calculus/syntax.py
class Expr:
    def __str__(self):
        raise NotImplementedError("Subclasses must implement __str__")
    
    def __eq__(self, other):
        raise NotImplementedError("Subclasses must implement __eq__")
    
    def free_variables(self):
        raise NotImplementedError("Subclasses must implement free_variables")
    
    def bound_variables(self):
        raise NotImplementedError("Subclasses must implement bound_variables")
    
    def substitute(self, var_name, replacement):
        raise NotImplementedError("Subclasses must implement substitute")
    
    def is_alpha_equivalent(self, other):
        raise NotImplementedError("Subclasses must implement is_alpha_equivalent")


class Variable(Expr):
    def __init__(self, name):
        self.name = name
    
    def __str__(self):
        return self.name
    
    def __eq__(self, other):
        if not isinstance(other, Variable):
            return False
        return self.name == other.name
    
    def free_variables(self):
        return {self.name}
    
    def bound_variables(self):
        return set()
    
    def substitute(self, var_name, replacement):

        if self.name == var_name:
            return replacement
        return self
    
    def is_alpha_equivalent(self, other):
        if not isinstance(other, Variable):
            return False
        return self.name == other.name


class Abstraction(Expr):
    def __init__(self, param, body):
        self.param = param
        self.body = body
    
    def __str__(self):
        return f"λ{self.param}.{str(self.body)}"
    
    def __eq__(self, other):
        if not isinstance(other, Abstraction):
            return False
        return self.param == other.param and self.body == other.body
    
    def free_variables(self):
        return self.body.free_variables() - {self.param}
    
    def bound_variables(self):
        return {self.param} | self.body.bound_variables()
    
    def substitute(self, var_name, replacement):
        
        if self.param == var_name:
            return self
        
        if self.param in replacement.free_variables():
            new_param = self._fresh_name(
                self.param, 
                replacement.free_variables() | self.body.free_variables()
            )
            new_body = self.body.substitute(self.param, Variable(new_param))
            final_body = new_body.substitute(var_name, replacement)
            return Abstraction(new_param, final_body)
        
        return Abstraction(self.param, self.body.substitute(var_name, replacement))
    
    def _fresh_name(self, base_name, used_names):
        i = 0
        new_name = base_name
        while new_name in used_names:
            i += 1
            new_name = f"{base_name}{i}"
        return new_name
    
    def is_alpha_equivalent(self, other):
        if not isinstance(other, Abstraction):
            return False
        
        if self.param == other.param:
            return self.body.is_alpha_equivalent(other.body)
        
        fresh_param = self._fresh_name(
            self.param, 
            self.body.free_variables() | other.body.free_variables()
        )
        
        self_body_renamed = self.body.substitute(self.param, Variable(fresh_param))
        other_body_renamed = other.body.substitute(other.param, Variable(fresh_param))
        
        return self_body_renamed.is_alpha_equivalent(other_body_renamed)


class Application(Expr):    
    def __init__(self, left, right):
        self.left = left
        self.right = right
    
    def __str__(self):
        left_str = f"({self.left})" if isinstance(self.left, Abstraction) else str(self.left)
        right_str = f"({self.right})" if isinstance(self.right, (Application, Abstraction)) else str(self.right)
        return f"{left_str} {right_str}"
    
    def __eq__(self, other):
        if not isinstance(other, Application):
            return False
        return self.left == other.left and self.right == other.right
    
    def free_variables(self):
        return self.left.free_variables() | self.right.free_variables()
    
    def bound_variables(self):
        return self.left.bound_variables() | self.right.bound_variables()
    
    def substitute(self, var_name, replacement):
        return Application(
            self.left.substitute(var_name, replacement),
            self.right.substitute(var_name, replacement)
        )
    
    def is_alpha_equivalent(self, other):
        if not isinstance(other, Application):
            return False
        return (self.left.is_alpha_equivalent(other.left) and 
                self.right.is_alpha_equivalent(other.right))

File: web/syntax.js
class Expr {
    toString() {
        throw new Error("Subclasses must implement toString");
    }
    
    freeVariables() {
        throw new Error("Subclasses must implement freeVariables");
    }

    boundVariables() {
        throw new Error("Subclasses must implement boundVariables");
    }
    substitute(varName, replacement) {
        throw new Error("Subclasses must implement substitute");
    }
    isAlphaEquivalent(other) {
        throw new Error("Subclasses must implement isAlphaEquivalent");
    }
}

class Variable extends Expr {
    constructor(name) {
        super();
        this.name = name;
    }
    
    toString() {
        return this.name;
    }
    
    freeVariables() {
        return new Set([this.name]);
    }
    
    boundVariables() {
        return new Set();
    }
    
    substitute(varName, replacement) {
        if (this.name === varName) {
            return replacement;
        }
        return this;
    }
    
    isAlphaEquivalent(other) {
        if (!(other instanceof Variable)) {
            return false;
        }
        return this.name === other.name;
    }
}

class Abstraction extends Expr {

    constructor(param, body) {
        super();
        this.param = param;
        this.body = body;
    }
    
    toString() {
        return `λ${this.param}.${this.body.toString()}`;
    }
    
    freeVariables() {
        const result = new Set(this.body.freeVariables());
        result.delete(this.param);
        return result;
    }
    
    boundVariables() {
        const result = new Set(this.body.boundVariables());
        result.add(this.param);
        return result;
    }
    

    _freshName(baseName, usedNames) {
        let i = 0;
        let newName = baseName;
        while (usedNames.has(newName)) {
            i++;
            newName = `${baseName}${i}`;
        }
        return newName;
    }
    
    substitute(varName, replacement) {
        if (this.param === varName) {
            return this;
        }
        
        if (replacement.freeVariables().has(this.param)) {
            const freeVarsUnion = new Set([
                ...replacement.freeVariables(),
                ...this.body.freeVariables()
            ]);
            const newParam = this._freshName(this.param, freeVarsUnion);
            
            const newBody = this.body.substitute(this.param, new Variable(newParam));
            
            const finalBody = newBody.substitute(varName, replacement);
            return new Abstraction(newParam, finalBody);
        }
        
        return new Abstraction(this.param, this.body.substitute(varName, replacement));
    }
    
    isAlphaEquivalent(other) {
        if (!(other instanceof Abstraction)) {
            return false;
        }
        
        if (this.param === other.param) {
            return this.body.isAlphaEquivalent(other.body);
        }
        
        const freeVarsUnion = new Set([
            ...this.body.freeVariables(),
            ...other.body.freeVariables()
        ]);
        const freshParam = this._freshName(this.param, freeVarsUnion);
        
        const selfBodyRenamed = this.body.substitute(this.param, new Variable(freshParam));
        const otherBodyRenamed = other.body.substitute(other.param, new Variable(freshParam));
        
        return selfBodyRenamed.isAlphaEquivalent(otherBodyRenamed);
    }
}

class Application extends Expr {

    constructor(left, right) {
        super();
        this.left = left;
        this.right = right;
    }
    
    toString() {
        const leftStr = this.left instanceof Abstraction ? 
            `(${this.left.toString()})` : this.left.toString();
        const rightStr = this.right instanceof Application || this.right instanceof Abstraction ? 
            `(${this.right.toString()})` : this.right.toString();
        return `${leftStr} ${rightStr}`;
    }
    
    freeVariables() {
        return new Set([
            ...this.left.freeVariables(),
            ...this.right.freeVariables()
        ]);
    }
    
    boundVariables() {
        return new Set([
            ...this.left.boundVariables(),
            ...this.right.boundVariables()
        ]);
    }
    
    substitute(varName, replacement) {
        return new Application(
            this.left.substitute(varName, replacement),
            this.right.substitute(varName, replacement)
        );
    }
    
    isAlphaEquivalent(other) {
        if (!(other instanceof Application)) {
            return false;
        }
        return (
            this.left.isAlphaEquivalent(other.left) && 
            this.right.isAlphaEquivalent(other.right)
        );
    }
}
